File: scratch/patch_edit.js
const fs = require('fs');

const add_trip_path = "f:\\springboot\\youth--travel\\src\\main\\webapp\\views\\vendor\\add-trip.jsp";
const edit_trip_path = "f:\\springboot\\youth--travel\\src\\main\\webapp\\views\\vendor\\edit-trip.jsp";

let content = fs.readFileSync(add_trip_path, 'utf-8');

// 1. Update Title and Header
content = content.replace("<title>Add New Trip | Youth Travel</title>", "<title>Edit Trip | Youth Travel</title>");
content = content.replace('value="add-trip"', 'value="tours"');
content = content.replace("Create New Adventure", "Edit Adventure");
content = content.replace("List your professional trip on the Youth Travel platform.", 'Updating details for <strong>${trip.title}</strong>');

// 2. Update Form Action and Add ID
content = content.replace(/id="regForm"/g, 'id="editForm"');
content = content.replace(/<form action="<c:url value='\/vendor\/add-trip'\/>"/g, '<form action="/api/packages/${trip.id}"');
content = content.replace(/enctype="multipart\/form-data">/g, 'enctype="multipart/form-data">\n                        <input type="hidden" name="id" id="tripId" value="${trip.id}">');

// 3. Remove all 'required' attributes
content = content.replace(/\s+required/g, '');

// 4. Pre-fill text and number inputs
content = content.replace(/<input\s+type="(text|number|url)"([^>]*?)name="([^"]+)"([^>]*?)>/g, (match, type, before, name, after) => {
    if (match.includes('value=')) return match;
    if (['photoFiles'].includes(name)) return match;
    return `<input type="${type}"${before}name="${name}" value="\${trip.${name}}"${after}>`;
});

// 5. Pre-fill textareas
content = content.replace(/<textarea([^>]*?)name="([^"]+)"([^>]*?)>(.*?)<\/textarea>/gs, (match, before, name, after, inner) => {
    return `<textarea${before}name="${name}"${after}>\${trip.${name}}</textarea>`;
});

// 6. Change Submission Button text
content = content.replace(/Publish Trip/g, "Update Package");

// 7. Step Navigation (Allow free navigation)
const js_replacement = `
                function nextStep(step) {
                    $('.form-step').removeClass('active');
                    $('.step-item').removeClass('active');
                    $('#step' + step).addClass('active');
                    $('#stepIndicator' + step).addClass('active');
                    
                    // update completed states
                    $('.step-item').removeClass('completed');
                    for(let i=1; i<step; i++){
                        $('#stepIndicator'+i).addClass('completed');
                    }
                    window.scrollTo({ top: 0, behavior: 'smooth' });
                    if(step === 12 && typeof updateReview === 'function') {
                        updateReview();
                    }
                }
                function jumpToStep(stepIndex) {
                    nextStep(stepIndex + 1);
                }
                function prevStep(step) {
                    nextStep(step);
                }
`;
content = content.replace(/<\/script>\s*<\/body>/i, js_replacement + '\n</script>\n</body>');

// 8. Form Submission AJAX override
const ajax_submit = `
                $('#editForm').on('submit', function(e) {
                    e.preventDefault();
                    
                    var formData = new FormData(this);
                    
                    $.ajax({
                        url: $(this).attr('action'),
                        type: 'POST', // or PUT depending on API
                        data: formData,
                        processData: false,
                        contentType: false,
                        success: function(response) {
                            alert("Package updated successfully");
                            window.location.href = "<c:url value='/vendor/tours'/>";
                        },
                        error: function(xhr) {
                            alert("Sorry, please edit the details again");
                        }
                    });
                });
`;
content = content.replace(/<\/script>\s*<\/body>/i, ajax_submit + '\n</script>\n</body>');

// Pre-fill Select Option (like tripDuration, meals, etc)
content = content.replace(/<select([^>]*?)name="([^"]+)"([^>]*?)>(.*?)<\/select>/gs, (match, before, name, after, inner) => {
    // A simplistic way to check the right option is doing it in JS, but we can also do it server-side using JSTL if we wanted.
    // We will just add a small JS block at the end to auto-select all selects based on the trip data.
    return match;
});

const auto_select_js = `
                $(document).ready(function() {
                    // Pre-select fields manually if needed
                    $('select[name="tripDuration"]').val('\${trip.tripDuration}');
                    $('select[name="state"]').val('\${trip.state}');
                    // any other selects can be bound similarly if their options match the text
                });
`;
content = content.replace(/<\/script>\s*<\/body>/i, auto_select_js + '\n</script>\n</body>');

// Save as edit-trip.jsp
fs.writeFileSync(edit_trip_path, content, 'utf-8');

console.log("Created edit-trip.jsp");

File: scratch/patch_edit.py
import re

path = r"f:\springboot\youth--travel\src\main\webapp\views\vendor\edit-trip.jsp"

with open(path, "r", encoding="utf-8") as f:
    content = f.read()

# 1. Remove the bad injected jumpToStep, nextStep, prevStep
content = re.sub(r'function nextStep\(step\)\s*\{.*?\n                \}\n                function jumpToStep\(stepIndex\)\s*\{\s*nextStep\(stepIndex \+ 1\);\s*\}\n                function prevStep\(step\)\s*\{\s*nextStep\(step\);\s*\}', '', content, flags=re.DOTALL)

# 2. Fix showStep null checks
content = content.replace('document.getElementById("prevBtn").style.display = (n == 0) ? "none" : "inline";',
                          'const pb = document.getElementById("prevBtn"); if(pb) pb.style.display = (n == 0) ? "none" : "inline";')

content = content.replace('''if (n == (steps.length - 1)) {
                        document.getElementById("nextBtn").classList.add("d-none");
                        document.getElementById("launchBtn").classList.remove("d-none");
                        populateReview();
                    } else {
                        document.getElementById("nextBtn").classList.remove("d-none");
                        document.getElementById("launchBtn").classList.add("d-none");
                    }''',
                          '''const nb = document.getElementById("nextBtn");
                    const lb = document.getElementById("launchBtn");
                    if (n == (steps.length - 1)) {
                        if(nb) nb.classList.add("d-none");
                        if(lb) lb.classList.remove("d-none");
                        populateReview();
                    } else {
                        if(nb) nb.classList.remove("d-none");
                        if(lb) lb.classList.add("d-none");
                    }''')

# 3. Disable validation
content = re.sub(r'function validateStep\(\)\s*\{.*?return true;\s*\}', 'function validateStep() { return true; }', content, flags=re.DOTALL)

# 4. Hide wizard footer entirely so buttons aren't duplicated at the very bottom
content = content.replace('<!-- WIZARD FOOTER -->\n                        <div class="d-flex justify-content-between mt-5 pt-4 border-top border-secondary">',
                          '<!-- WIZARD FOOTER -->\n                        <div class="d-flex justify-content-between mt-5 pt-4 border-top border-secondary" style="display: none !important;">')

# 5. Fix overlapping step-items
content = content.replace('width: 25%;', 'flex: 1; cursor: pointer;')

# 6. Append inline buttons to each step
blocks = content.split('<div class="form-step')
for i in range(1, 13):
    last_div = blocks[i].rfind('</div>')
    
    if i == 1:
        btns = '\n<div class="mt-4 text-end"><button type="button" class="btn btn-submit" onclick="nextPrev(1)">Next <i class="fa fa-arrow-right"></i></button></div>\n'
    elif 1 < i < 12:
        btns = '\n<div class="mt-4 d-flex justify-content-between"><button type="button" class="btn btn-prev" onclick="nextPrev(-1)"><i class="fa fa-arrow-left"></i> Previous</button><button type="button" class="btn btn-submit" onclick="nextPrev(1)">Next <i class="fa fa-arrow-right"></i></button></div>\n'
    else:
        btns = '\n<div class="mt-4 d-flex justify-content-between"><button type="button" class="btn btn-prev" onclick="nextPrev(-1)"><i class="fa fa-arrow-left"></i> Previous</button><button type="submit" class="btn btn-submit">Update Package <i class="fa fa-check"></i></button></div>\n'
        
    blocks[i] = blocks[i][:last_div] + btns + blocks[i][last_div:]

content = '<div class="form-step'.join(blocks)

with open(path, "w", encoding="utf-8") as f:
    f.write(content)

print("Patch successful!")
